# Bioinformatic armory solutions

As they mention, in this section, the Armory, we will familiarize ourselves with a sampling of some of the more popular bioinformatics tools taken from "out of the box" software. To be equitable, we will focus mainly on free, internet-based software and on programs that are compatible with multiple operating systems. The "Problem" section will contain links to this software, with short descriptions about how to use it.

The very first basic problem is to count DNA nucleotides. Find additional information in the rosalind website link [here](https://rosalind.info/problems/ini/).

Find previous information and example of solutions in the Stronghold solutions interactive python notebook here [`../Stronghold/stronghold_solutions.ipynb`](../Stronghold/stronghold_solutions.ipynb)

In [ ]:
dna_string = "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC"

Now, try to be more ellaborate and produce a more programmatic solution

A very pythonic solution...

In [ ]:
## Solution 1: Using a python string
nucs = ["A", "C", "G", "T"]
counts = [dna_string.count(nuc) for nuc in nucs]
final_string = " ".join(map(str, counts))
print(final_string)


In [ ]:
## solution 2: Using biopython seq
from Bio.Seq import Seq
my_seq = Seq(dna_string)

nuc_list = [my_seq.count(nuc) for nuc in nucs]
final_string = " ".join(map(str, nuc_list))
print(final_string)

Create a single file execution script from command line. See file in [`code/count_nuc.py`](./code/count_nuc.py)

## GenBank Introduction

The following exercise is about producing a NCBI Genbank search using python. See details [here](https://rosalind.info/problems/gbk/).


For details and examples for how Biopython works, take a look at the [Example_Biopython.ipynb](https://github.com/JFsanchezherrero/Python_course/blob/main/practical_exercises/biopython/Example_Biopython.ipynb) available within my Python_course [here](https://github.com/JFsanchezherrero/Python_course/tree/main).

The goal of this exercise is to retrieve the number of nucleotide genbank entries for a given genus name published between the dates specified in YYYY/M/D format.

In [36]:
# as an example
information2use = '''Anthoxanthum
2003/7/25
2005/12/27'''

In [37]:
info_list = information2use.split("\n")

Find details of all limits that can be applied in Genbank searches [here](https://www.ncbi.nlm.nih.gov/nuccore/limits)

### Filter by genus

First, create a simple search using just the genus name

In [38]:
string2search = info_list[0] + "[ORGN]"

print(string2search)

Anthoxanthum[ORGN]


In [39]:
# load Biopython module Entrez
from Bio import Entrez
Entrez.email = 'JoseFrancisco.Sanchez@uab.cat'

#Connect GenBank:
handle = Entrez.esearch(db="nucleotide", term=string2search)
## provide the term of interest using appropiate tags [ORGN], [TITL]

# Save results (XML format as default)
# and close connection
results = Entrez.read(handle)
handle.close()

What type of object it is? 

In [40]:
type(results)

Bio.Entrez.Parser.DictionaryElement

It is a special Biopython dictionary and we can access keys, values and items

In [41]:
results.items()

dict_items([('Count', '85458'), ('RetMax', '20'), ('RetStart', '0'), ('IdList', ['3351085915', '3345836397', '3345836321', '3345836169', '3326832244', '3326832207', '3326832140', '3037294572', '3037294485', '3038019637', '3038019631', '3038019565', '3038019559', '3010878933', '3010878932', '2989059421', '2954372271', '2634197321', '2553703756', '2511466363']), ('TranslationSet', [{'From': 'Anthoxanthum[ORGN]', 'To': '"Anthoxanthum"[Organism]'}]), ('TranslationStack', [{'Term': '"Anthoxanthum"[Organism]', 'Field': 'Organism', 'Count': '85458', 'Explode': 'Y'}, 'GROUP']), ('QueryTranslation', '"Anthoxanthum"[Organism]')])

In [42]:
results.values()

dict_values(['85458', '20', '0', ['3351085915', '3345836397', '3345836321', '3345836169', '3326832244', '3326832207', '3326832140', '3037294572', '3037294485', '3038019637', '3038019631', '3038019565', '3038019559', '3010878933', '3010878932', '2989059421', '2954372271', '2634197321', '2553703756', '2511466363'], [{'From': 'Anthoxanthum[ORGN]', 'To': '"Anthoxanthum"[Organism]'}], [{'Term': '"Anthoxanthum"[Organism]', 'Field': 'Organism', 'Count': '85458', 'Explode': 'Y'}, 'GROUP'], '"Anthoxanthum"[Organism]'])

In [43]:
results.keys()

dict_keys(['Count', 'RetMax', 'RetStart', 'IdList', 'TranslationSet', 'TranslationStack', 'QueryTranslation'])

We can inspect some of the results

In [44]:
results['Count']

'85458'

In [45]:
results['RetMax']

'20'

There were a total of 85458 hits in database, only the first 20 available in IdList

In [46]:
results["IdList"]

['3351085915', '3345836397', '3345836321', '3345836169', '3326832244', '3326832207', '3326832140', '3037294572', '3037294485', '3038019637', '3038019631', '3038019565', '3038019559', '3010878933', '3010878932', '2989059421', '2954372271', '2634197321', '2553703756', '2511466363']

### Filter by genus and date 

For organism we used the key tag [ORGN]; for date, we have to use [PDAT] that accounts for publication date. See an example for our previous case.

In [51]:
string2search = info_list[0] + "[ORGN] AND (\"" + info_list[1] + "\"[PDAT]:\"" + info_list[2] + "\"[PDAT])"

print(string2search)

Anthoxanthum[ORGN] AND ("2003/7/25"[PDAT]:"2005/12/27"[PDAT])


In [54]:
# load Biopython module Entrez
from Bio import Entrez
Entrez.email = 'JoseFrancisco.Sanchez@uab.cat'

#Connect GenBank:
handle = Entrez.esearch(db="nucleotide", term=string2search)
## provide the term of interest using appropiate tags [ORGN], [TITL]

# Save results (XML format as default)
# and close connection
results = Entrez.read(handle)
handle.close()

This list of items is a list of entries in Genbank database. As an example, the first entry (57283843) is https://www.ncbi.nlm.nih.gov/nuccore/57283843

In [55]:
results

{'Count': '11', 'RetMax': '11', 'RetStart': '0', 'IdList': ['51493519', '52631901', '52631900', '52631899', '33413983', '33413982', '33413981', '33413980', '71056585', '57283843', '57283791'], 'TranslationSet': [{'From': 'Anthoxanthum[ORGN]', 'To': '"Anthoxanthum"[Organism]'}], 'TranslationStack': [{'Term': '"Anthoxanthum"[Organism]', 'Field': 'Organism', 'Count': '85458', 'Explode': 'Y'}, {'Term': '"2003/7/25"[PDAT]', 'Field': 'PDAT', 'Count': '0', 'Explode': 'N'}, {'Term': '"2005/12/27"[PDAT]', 'Field': 'PDAT', 'Count': '0', 'Explode': 'N'}, 'RANGE', 'GROUP', 'AND'], 'QueryTranslation': '"Anthoxanthum"[Organism] AND ("2003/7/25"[PDAT] : "2005/12/27"[PDAT])'}

Create a single file execution script from command line. See file in [`code/genbanck_search.py`](./code/genbanck_search.py)